aiaccel/hpo/samplers/nelder_mead_sampler.py

```diff
@@ -0,0 +1,191 @@
+
+    def __iter__(self) -> Generator[np.ndarray[Any, Any], None, None]:
+        # initialization
+        lows, highs = zip(*self._search_space.values())
+        self.vertices = self._rng.uniform(lows, highs, (self.dimension + 1, self.dimension))
+
+        self.is_ready = True
+        yield from iter(self.vertices[:-1])
+        self.is_ready = False
+        yield self.vertices[-1]
+        self.values = np.array([self.value_queue.get() for _ in range(len(self.vertices))])
+
+        # main loop
+        shrink_requied = False
+        while True:
+            # sort self.vertices by their self.values
+            order = np.argsort(self.values)
+            self.vertices, self.values = self.vertices[order], self.values[order]
+
+            # reflect
+            yc = self.vertices[:-1].mean(axis=0)
+            yield (yr := yc + self.coeff.r * (yc - self.vertices[-1]))
+
+            fr = self.value_queue.get()
+
+            if self.values[0] <= fr < self.values[-2]:
+                self.vertices[-1], self.values[-1] = yr, fr
+            elif fr < self.values[0]:  # expand
+                yield (ye := yc + self.coeff.e * (yc - self.vertices[-1]))
+
+                fe = self.value_queue.get()
+
+                self.vertices[-1], self.values[-1] = (ye, fe) if fe < fr else (yr, fr)
+            elif self.values[-2] <= fr < self.values[-1]:  # outside contract
+                yield (yoc := yc + self.coeff.oc * (yc - self.vertices[-1]))
+                foc = self.value_queue.get()
+
+                if foc <= fr:
+                    self.vertices[-1], self.values[-1] = yoc, foc
+                else:
+                    shrink_requied = True
+            elif self.values[-1] <= fr:  # inside contract
+                yield (yic := yc + self.coeff.ic * (yc - self.vertices[-1]))
+                fic = self.value_queue.get()
+
+                if fic < self.values[-1]:
+                    self.vertices[-1], self.values[-1] = yic, fic
+                else:
+                    shrink_requied = True
+
+            # shrink
+            if shrink_requied:
+                self.vertices = self.vertices[0] + self.coeff.s * (self.vertices - self.vertices[0])
+                self.is_ready = True
+                yield from iter(self.vertices[1:-1])
+                self.is_ready = False
+                yield self.vertices[-1]
+
+                self.values[1:] = [self.value_queue.get() for _ in range(len(self.vertices) - 1)]
+
+                shrink_requied = False
+
+
+class NelderMeadSampler(optuna.samplers.BaseSampler):
+    def __init__(
+        self,
+        search_space: dict[str, list[float]],
+        seed: int | None = None,
+        coeff: NelderMeadCoefficient | None = None,
+    ) -> None:
+        self._search_space = {}  # Memorise parameter order.
+        for param_name, param_distribution in sorted(search_space.items()):
+            self._search_space[param_name] = list(param_distribution)
```

Comment:
```suggestion
        self._search_space = {name: list(dist) for name, dist in search_space.items()}
```
と書くとすっきりできそうです．Python3.7からdictはordered dictなので，sortは不要かと思いました．